packages/dates/src/format.ts

```diff
@@ -0,0 +1,223 @@
+  | 'H'
+  | 'hh'
+  | 'h'
+  | 'mm'
+  | 'm'
+  | 'ss'
+  | 's'
+  | 'A'
+  | 'a'
+  | 'dddd'
+  | 'ddd';
+
+type TokenDateTimeFormatOptionsMap = {
+  [key in Token]: {
+    type: Intl.DateTimeFormatPartTypes;
+    options: DateTimeFormatOptions;
+    extraParser?: (str: string) => string;
+  };
+};
+
+const TOKEN_MAP: TokenDateTimeFormatOptionsMap = {
+  YYYY: {type: 'year', options: {year: 'numeric'}},
+  YY: {type: 'year', options: {year: '2-digit'}},
+  MMMM: {type: 'month', options: {month: 'long'}},
+  MMM: {type: 'month', options: {month: 'short'}},
+  MM: {type: 'month', options: {month: '2-digit'}},
+  // eslint-disable-next-line id-length
+  M: {type: 'month', options: {month: 'numeric'}},
+  DD: {type: 'day', options: {day: '2-digit'}},
+  // eslint-disable-next-line id-length
+  D: {type: 'day', options: {day: 'numeric'}},
+  HH: {type: 'hour', options: {hour: '2-digit', hour12: false}},
+  // eslint-disable-next-line id-length
+  H: {
+    type: 'hour',
+    options: {hour: 'numeric', hour12: false},
+    // extraParser is needed when Node and Browsers don't respect the 2-digit
+    // and numeric options, which happens occasionally for time units
+    extraParser: numericFromTwoDigit,
+  },
+  hh: {type: 'hour', options: {hour: '2-digit', hour12: true}},
+  // eslint-disable-next-line id-length
+  h: {type: 'hour', options: {hour: 'numeric', hour12: true}},
+  mm: {
+    type: 'minute',
+    options: {minute: '2-digit'},
+    // see 'H' token parsing for more details about extraParser
+    extraParser: twoDigitFromNumeric,
+  },
+  // eslint-disable-next-line id-length
+  m: {type: 'minute', options: {minute: 'numeric'}},
+  ss: {
+    type: 'second',
+    options: {second: '2-digit'},
+    extraParser: twoDigitFromNumeric,
+  },
+  // eslint-disable-next-line id-length
+  s: {type: 'second', options: {second: 'numeric'}},
+  // eslint-disable-next-line id-length
+  A: {
+    type: 'dayPeriod',
+    options: {
+      hour: '2-digit',
+      minute: '2-digit',
+      second: '2-digit',
+      hour12: true,
+    },
+  },
+  // eslint-disable-next-line id-length
+  a: {
+    type: 'dayPeriod',
+    options: {
+      hour: '2-digit',
+      minute: '2-digit',
+      second: '2-digit',
+      hour12: true,
+    },
+    extraParser: lowercase,
+  },
+  dddd: {type: 'weekday', options: {weekday: 'long'}},
+  ddd: {type: 'weekday', options: {weekday: 'short'}},
+};
+
+function numericFromTwoDigit(time: string) {
+  return time.startsWith('0') && time.length === 2 ? time.charAt(1) : time;
+}
+
+function twoDigitFromNumeric(time: string) {
+  return time.length === 1 ? `0${time}` : time;
+}
+
+function lowercase(time: string) {
+  return time.toLowerCase();
+}
+
+export function format(
+  date: Date,
+  fmtStr: string,
+  tz?: string,
+  localeStr?: string,
+): string {
+  const {locale, timeZone} = memoizedGetDateTimeFormat(localeStr, {
+    timeZone: tz,
+  }).resolvedOptions();
+  let result = fmtStr;
+
+  // the strategy is as follows:
```

Comment:
Couldn't this have been done with one pass of `fmtStr.replace(GIANT_REGEXP_OF_ALL_TOKENS, () => ...)`?

Reply:
I think, yes it could be done this way. In this circumstance, the callback would go through each match and make the replacement as specified by `TOKEN_MAP`, and then we wouldn't have to do multiple regex passes or surrounding the replacements with braces.

I'll try it out and see! Thanks for suggesting, I hadn't thought of it that way. 👍

Reply:
Also, I'd be happy to pair on it!

Reply:
Yeah I can take you up on that!